File: messaging/broker.factory.ts
import { MessageBroker } from './broker.interface';
import { RabbitMQBroker } from './rabbitmq.broker';
import { SQSBroker } from './sqs.broker';

export function createBroker(): MessageBroker {
  const type = process.env.BROKER_TYPE ?? 'rabbitmq';

  switch (type) {
    case 'sqs':
      return new SQSBroker(process.env.AWS_REGION ?? 'us-east-1');
    case 'rabbitmq':
      return new RabbitMQBroker(process.env.RABBITMQ_URL ?? 'amqp://localhost');
    default:
      throw new Error(`Unknown BROKER_TYPE: "${type}"`);
  }
}

File: src/sourcing/messaging/broker_factory.py
"""Factory for broker implementations based on configuration."""

from __future__ import annotations

import os
from pathlib import Path

from dotenv import load_dotenv

from messaging.broker_interface import BrokerInterface
from messaging.rabbitmq_broker import RabbitmqBroker

_SOURCING_ROOT = Path(__file__).resolve().parent.parent


def create_broker() -> BrokerInterface:
    """Return a broker instance for ``BROKER_TYPE`` (default: ``rabbitmq``)."""
    # Local runs: load src/sourcing/.env so RABBITMQ_URL etc. match Docker env_file behavior.
    load_dotenv(_SOURCING_ROOT / ".env", override=False)

    broker_type = os.environ.get("BROKER_TYPE", "rabbitmq").strip().lower()
    if broker_type == "rabbitmq":
        return RabbitmqBroker()
    raise ValueError(f"Unsupported BROKER_TYPE: {broker_type!r}. Use 'rabbitmq'.")
